### ISSUE TITLE
Build failures with CMake and Visual Studio

### ISSUE THREAD

**Issue by arjenmarkus** [FluidNumerics/feq-parse#14]
I experienced a few problems when I tried to build the library and the examples via Visual Studio and Intel Fortran oneAPI on Windows. The steps I took are simple enough:

- Run CMake with default settings. On my system I get a VS solution for Visual Studio 17 2022.
- Build solution: the first round gives a build problem regarding the module file FEQPARSE in the test programs. 
- Building it again only gives a single error: lib ..\src\libfeqparse-static.a does not exist. This is for "testsuite"
- Because the other problems are gone I suspect that the test programs are not properly dependent on the feqparse library. Indeed "testsuite" does not depend on anything but "ZERO_CHECK"(a CMake target).
- In tests/CMakeLists.txt I see that a fixed name is used for the libfeqparse-static library. On Windows this would have the extension .lib.

**Comment by fluidnumerics-joe** [comment]
Thanks for reporting. I don't have access to a Windows system with visual studio. If you have modifications for the CMake build system that function for Windows with visual studio and the Intel compilers, feel free to open a pull request.

**Comment by arjenmarkus** [comment]
Hi, well, that should be no particular problem. Will do.

Op za 2 dec 2023 om 15:58 schreef Joe Schoonover ***@***.***>:

**Comment by arjenmarkus** [comment]
Well, the linkage problem is gone - that was easy, but the compile problems
are confounding:
There are four source files in the test directory that present problems:

Build Log


 Build started: Project: testsuite_fortran, Configuration: Debug|x64
Output
      Compiling with Intel® Fortran Compiler Classic 2021.9.0 [Intel(R)
64]...
ifort /nologo /debug:full /Od /fpp /I"D:\fortran\feq-parse\build\include"
/I"D:\fortran\feq-parse\build\include/Debug" /DCMAKE_INTDIR=\"Debug\"
/module:"..\include\Debug" /object:"testsuite_fortran.dir\Debug\\"
/Fd"testsuite_fortran.dir\Debug\vc170.pdb" /libs:dll /threads /dbglibs /c
/W1 /Qlocation,link,"C:\Program Files\Microsoft Visual
Studio\2022\Professional\VC\Tools\MSVC\14.36.32532\bin\HostX64\x64" /Qm64
"D:\fortran\feq-parse\test\linear_r1fp32.f90"
D:\fortran\feq-parse\test\linear_r1fp32.f90(3): error #7002: Error in
opening the compiled module file.  Check INCLUDE paths.   [FEQPARSE]
  USE FEQParse
------^
D:\fortran\feq-parse\test\linear_r1fp32.f90(7): error #6457: This derived
type name has not been declared.   [EQUATIONPARSER]
  TYPE(EquationParser) :: f
-------^
D:\fortran\feq-parse\test\linear_r1fp32.f90(22): error #6404: This name
does not have a type, and must have an explicit type.   [F]
    f = EquationParser(eqChar, independentVars)
----^
compilation aborted for D:\fortran\feq-parse\test\linear_r1fp32.f90 (code 1)


The files in question:

   - linear_r1fp32.f90
   - asin_r1fp32.f90
   - atan_sfp64.f90
   - tan_sfp32.f90

All others are fine. It is not a problem with parallel builds, as the list
is always the same.

I have compared the generated command-lines for one of the above files with
one that does get compiled the first time around and there is no difference
in the options.

I am puzzled :(.


Op zo 3 dec 2023 om 11:09 schreef Arjen Markus ***@***.***>:

**Comment by arjenmarkus** [comment]
Actually, yes and the other files are compiled without any complaint
whatsoever. I will ask on Discourse whether someone has a solution.

BTW, the fix for the feqparse-static library is:

    target_link_libraries (${TARGET} ${TARGET}_fortran feqparse-static)

and remove the line:

    target_link_libraries (${TARGET}
${CMAKE_BINARY_DIR}/src/libfeqparse-static.a)

(I Have not figured out how to use the variable holding the name, but this
works)

Op zo 3 dec 2023 om 21:27 schreef Joe Schoonover ***@***.***>:

**Comment by fluidnumerics-joe** [comment]
@arjenmarkus - I added another commit to the [bugfix/issue-12 branch](https://github.com/FluidNumerics/feq-parse/tree/bugfix/issue-12) that I *think* will resolve the issue.

What seems to be happening is that the `${TARGET}_fortran` target is building before the `feqparse-static` target is finished building. The `test/CMakeLists.txt` didn't really have an explicit dependency between compiling the `src/` directory and compiling the `.f90` files in the test directory (though there is a dependency between compiling the generated `main.c` and `feqparse-static`). I've added an `add_dependencies` statement to force `feqparse-static` to be built entirely before building any of the `test/*.f90` files. I'd be interested to see if this resolves your CMake issue.

I've configured VSCode on Ubuntu 22.04 to use the CMake build system and it seems to build just fine

**Comment by arjenmarkus** [comment]
Right, "add_dependencies" is the statement I was looking for! That did the
trick, it was all built in one go, no issues.

I ran the test program:

   - The option -A does not seem to work for some reason, at least I see no
   output at all.
   - Several tests, 131 and 132 in particular (I did not manually try all
   of them), give a long list of warnings - Attempt to pop from empty token
   stack.

Other than that the build process runs smoothly :).



Op ma 4 dec 2023 om 20:17 schreef Joe Schoonover ***@***.***>:
